tensorflow/compiler/mlir/hlo/lib/Analysis/userange_analysis.cc

```diff
@@ -291,6 +285,47 @@ void UseInterval::intervalSubtract(UseInterval::Vector &a,
   }
 }
 
+/// Performs an interval intersection => A = A ^ B.
+void UseInterval::intervalIntersect(UseInterval::Vector &a,
+                                    const UseInterval::Vector &b) {
+  auto iterB = b.begin();
+  auto endB = b.end();
+  for (auto iterA = a.begin(); iterA != a.end() && iterB != endB;) {
+    // iterA is strictly before iterB => erase iterA.
+    if (*iterA < *iterB) {
+      iterA = a.erase(iterA);
+      // iterB is strictly before iterA => increment iterB.
+    } else if (*iterA > *iterB) {
+      ++iterB;
+      // iterA overlaps with the start of iterB. Set the start of iterA to the
+      // start of iterB and increment iterA.
+    } else if (iterA->start < iterB->start && iterA->end <= iterB->end) {
+      iterA->start = iterB->start;
+      ++iterA;
+      // iterA overlaps with the end of iterB.
+    } else if (iterA->end > iterB->end) {
+      // If iterA overlaps with the start of iterB, the start of iterA must be
+      // set to the start of iterB.
+      if (iterA->start < iterB->start)
```

Comment:
You could also say that `iterA->start = std::max(iterA->start, iterB->start)`. That might convey better what is happening here.